--- google_appengine/google/appengine/dist/_library.py
+++ google_appengine/google/appengine/dist/_library.py
@@ -49,7 +49,7 @@ import os
 import sys
 
 server_software = os.getenv('SERVER_SOFTWARE')
-USING_SDK = not server_software or server_software.startswith('Dev')
+USING_SDK = True #not server_software or server_software.startswith('Dev')
 del server_software


--- a/lib/fancy_urllib/fancy_urllib/__init__.py
+++ b/lib/fancy_urllib/fancy_urllib/__init__.py
@@ -85,7 +85,7 @@ def create_fancy_connection(tunnel_host=None, key_file=None,
       self.ca_certs = ca_certs
       if can_validate_certs():
         if self.ca_certs:
-          self.cert_reqs = ssl.CERT_REQUIRED
+          self.cert_reqs = ssl.CERT_NONE #ssl.CERT_REQUIRED
         else:
           self.cert_reqs = ssl.CERT_NONE
 
